psutil/arch/osx/cpu.c

```diff
@@ -109,6 +112,48 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
     );
 }
 
+PyObject *
+psutil_arm_cpu_freq(PyObject *self, PyObject *args) {
+    unsigned int curr;
+    uint32_t min = UINT32_MAX;
+    uint32_t max = 0;
+
+    CFDictionaryRef matching = IOServiceMatching("AppleARMIODevice");
+    io_iterator_t iter;
+
+    IOServiceGetMatchingServices(kIOMainPortDefault, matching, &iter);
+    io_registry_entry_t entry;
+    while ((entry = IOIteratorNext(iter))) {
+        io_name_t name;
+        IORegistryEntryGetName(entry, name);
+
+        if (strncmp(name, "pmgr", 4) == 0) {
+            break;
+        }
+    }
+    IOObjectRelease(iter);
+    CFRelease(matching);
+    CFTypeRef pCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states5-sram"), kCFAllocatorDefault, 0);
+
+    size_t length = CFDataGetLength(pCoreRef);
+    for (size_t i = 0; i < length - 3; i += 4) {
```

Comment:
While this iteration works for this particular use case, it's inaccurate.  The array you are iterating over contains pairs of (4-byte) integers; the first one is the frequency you're looking for (little-endian).   For example, here's the array you're parsing from an [online M1 dump](https://github.com/xzhih/one-key-hidpi/files/5612845/ioreg_output_m1mac.txt), split into the pairs of numbers:
```
0046c323 0d030000
00475a31 0d030000
0048f13e 0d030000
0049884c 20030000
002f6859 2c030000
0030ff66 3f030000
00319674 61030000
00322d82 8d030000
00fd558e b9030000
00c87e9a eb030000
005d39a5 1d040000
004319b2 39040000
003689b8 39040000
00a265bb 39040000
0029f9be 39040000
```
The min value you want is 0x23c34600, the first integer in the array.
The max value you want is 0xbef92900, the second-to-last integer in the array.

Here's a screenshot of my M2 Max E-core values:
<img width="733" alt="Screenshot 2023-04-08 at 11 08 05 AM" src="https://user-images.githubusercontent.com/9291703/230736488-9a28eb12-99ec-41db-b7e2-e4840946ee55.png">

My min E-core freq is 0x365c0400 (912 MHz).
My max E-core freq is 0x907b4e00 (2.424 GHz).

And the P-core values:
<img width="736" alt="Screenshot 2023-04-08 at 11 16 48 AM" src="https://user-images.githubusercontent.com/9291703/230736850-6f39e7c5-9f28-432c-9e09-ed704ec6384f.png">

My min P-core freq is 0x29d7ab80 (702 MHz).
My max P-core freq is 0xdc4c7c00 (3.696 GHz).

So you could (probably) grab from index 0 and index (length - 8) and avoid iteration.  However, I'm only guessing that the array is always sorted in increasing order, as I haven't found that documented anywhere but have confirmed it with every sample output I could get my hands on.  If you do iterate, I would increment by 8 rather than 4.  That will save you the need to do your `curr_freq > 1e6` test below.

Reply:
Hey Daniel

> Compare to powermetrics output with my CPUs idling (mostly) at min:

It's interesting to note that `powermetrics` output shows it's able to return per-CPU frequencies. Current psutil's [cpu_freq() implementation](https://github.com/giampaolo/psutil/blob/7eadee31db2f038763a3a6f978db1ea76bbc4674/psutil/arch/osx/cpu.c#L113-L140) on OSX doesn't do that (same on Windows and OpenBSD).

Reply:
> It's interesting to note that `powermetrics` output shows it's able to return per-CPU frequencies. Current psutil's [cpu_freq() implementation](https://github.com/giampaolo/psutil/blob/7eadee31db2f038763a3a6f978db1ea76bbc4674/psutil/arch/osx/cpu.c#L113-L140) on OSX doesn't do that (same on Windows and OpenBSD).

Yeah, I've been chasing the "current frequency" nut for years.  I've always thought that CPU frequency was something like Heisenberg's uncertainty principle.  You change it by measuring it. :)

`powermetrics` is useful but is also closed-source so I don't know how it works, but it isn't really suitable for command-line parsing:
 - it polls every 5 seconds 
 - it requires sudo
 - it has differing output per-core.

That said, I will show you how powermetrics gets its "current frequency" and try to amplify your comment on your review,
> I am not sure I understand what is going on here. It seems it's based too much on guessing.

I'd prefer to use the term "reverse engineering". :) There are plenty of ARM CPU IORegistry dumps out there to validate these particular calls from iPhones, and I've personally tested them on both M1 and M2.

The pairs of numbers are most definitely frequency values.  I just wrote a quick program to print out the pairs on my M2:
```
voltage-states5-sram
702000000 / 790
948000000 / 790
1188000000 / 790
1452000000 / 806
1704000000 / 834
1968000000 / 859
2208000000 / 896
2400000000 / 915
2568000000 / 931
2724000000 / 978
2868000000 / 1009
3000000000 / 1062
3132000000 / 1087
3264000000 / 1140
3360000000 / 1140
3408000000 / 1140
3504000000 / 1140
3528000000 / 1196
3696000000 / 1196

voltage-states1-sram
912000000 / 590
1284000000 / 625
1752000000 / 678
2004000000 / 706
2256000000 / 778
2424000000 / 831
```
As I mentioned before I'm not sure what the second number is, but the first one is definitely frequency, as the more detailed `powermetrics` output shows. It gives percentage of time (probably over that 5-second polling interval) that the CPU was operating at each of its possible speeds.

CPU 0 (an E-core) spent 100% of its time in the lowest 912 MHz setting; but you can see the other values exactly line up with the voltage states IOReg value.
```
CPU 0 frequency: 912 MHz
CPU 0 idle residency:   0.00%
CPU 0 active residency: 100.00% (912 MHz: 100% 1284 MHz: .00% 1752 MHz:   0% 2004 MHz:   0% 2256 MHz:   0% 2424 MHz:   0%)
```
CPU 4 (a P-core) is a bit more insteresting, the 751 MHz "current frequency" is a weighted average of the possible frequencies by the amount of time spent in them:
```
CPU 4 frequency: 751 MHz
CPU 4 idle residency:  15.76%
CPU 4 active residency:  84.24% (702 MHz:  68% 948 MHz:  16% 1188 MHz:   0% 1452 MHz: .03% 1704 MHz: .00% 1968 MHz: .13% 2208 MHz: .03% 2400 MHz: .01% 2568 MHz:   0% 2724 MHz:   0% 2868 MHz:   0% 3000 MHz:   0% 3132 MHz:   0% 3264 MHz:   0% 3360 MHz: .00% 3408 MHz:   0% 3504 MHz:   0% 3528 MHz:   0% 3696 MHz:   0%)
```
Note that 751 MHz is only an average for the 84.24% of time that it was active.  What was its frequency when idle?  Zero?  So perhaps the average should really be 629 MHz?   

TLDR: "Current Frequency" on any CPU is too poorly defined to really measure.  However, these outputs show these voltage states as a reliable source for the possible settings.

Is it documented these outputs are always in ascending order?  No, I'm guessing. But it would make no sense for outputs like the above to have to sort each time, so it's a reasonable guess that every other 4 bytes is a possible frequency, so index 0 is the min frequency and index (length-8) is the max frequency for that core.

Reply:
This is an interesting search: https://github.com/search?q=voltage-states5-sram&type=code

It includes:
 - My project (oshi) and a project BUS which seems to copy my (MIT) code :) 
 - The go project @shoenig mentioned in the issue
 - qemu which hardcodes those voltage states arrays when emulating a t8030 chip (A13)
 - an `osquery` project with [code that looks a lot like this PR](https://github.com/cdprf/osquery/blob/aea0d6ef30a5e38f66cca252ebfcebe80d1f231a/osquery/tables/system/darwin/cpu_info.cpp#L145-L155)

Reply:
My current working theory is the second number in each pair is the voltage (in millivolts) that is set when the corresponding frequency is in use (e.g., 1.2V at highest frequencies).  Would make sense given the name :)

Reply:
> Also I notice you're only iterating the P-core values and not the E-core ones. Is this intentional?

I was under the impression that E-core values have a higher minimum frequency and lower max frequency than P-core ones. However, I think only the latter statement is true. So to find the minimum frequency you would also need to query the E-core values as well and compare them against the others.

Reply:
> I was under the impression that E-core values have a higher minimum frequency and lower max frequency than P-core ones. 

That's true now, but I'm not sure it will always be true.